contracts/Proofs.sol

```diff
@@ -1,123 +1,98 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.8;
 
+import "./Configuration.sol";
 import "./Requests.sol";
 import "./Periods.sol";
 
 abstract contract Proofs is Periods {
-  uint256 private immutable timeout;
-  uint8 private immutable downtime;
-
-  constructor(
-    uint256 __period,
-    uint256 __timeout,
-    uint8 __downtime
-  ) Periods(__period) {
+  ProofConfig private config;
+
+  constructor(ProofConfig memory _config) Periods(_config.period) {
     require(block.number > 256, "Insufficient block height");
-    timeout = __timeout;
-    downtime = __downtime;
+    config = _config;
   }
 
-  mapping(SlotId => bool) private slotIds;
   mapping(SlotId => uint256) private slotStarts;
   mapping(SlotId => uint256) private probabilities;
   mapping(SlotId => uint256) private missed;
   mapping(SlotId => mapping(Period => bool)) private received;
   mapping(SlotId => mapping(Period => bool)) private missing;
 
-  function _timeout() internal view returns (uint256) {
-    return timeout;
-  }
-
-  // Override this to let the proving system know when proofs for a
-  // slot are no longer required.
-  function proofEnd(SlotId id) public view virtual returns (uint256);
+  function slotState(SlotId id) public view virtual returns (SlotState);
 
   function missingProofs(SlotId slotId) public view returns (uint256) {
     return missed[slotId];
   }
 
-  /// @notice Informs the contract that proofs should be expected for id
-  /// @dev Requires that the id is not already in use
-  /// @param probability The probability that a proof should be expected
   function _startRequiringProofs(SlotId id, uint256 probability) internal {
-    require(!slotIds[id], "Proofs already required for slot");
-    slotIds[id] = true;
     slotStarts[id] = block.timestamp;
     probabilities[id] = probability;
   }
 
-  function _stopRequiringProofs(SlotId id) internal {
-    require(slotIds[id], "Proofs not required for slot");
-    slotIds[id] = false;
-  }
-
-  function _getPointer(
-    SlotId id,
-    Period proofPeriod
-  ) internal view returns (uint8) {
+  function getPointer(SlotId id, Period period) internal view returns (uint8) {
     uint256 blockNumber = block.number % 256;
-    uint256 periodNumber = Period.unwrap(proofPeriod) % 256;
+    uint256 periodNumber = Period.unwrap(period) % 256;
     uint256 idOffset = uint256(SlotId.unwrap(id)) % 256;
     uint256 pointer = (blockNumber + periodNumber + idOffset) % 256;
     return uint8(pointer);
   }
 
-  function _getPointer(SlotId id) internal view returns (uint8) {
-    return _getPointer(id, blockPeriod());
+  function getPointer(SlotId id) public view returns (uint8) {
+    return getPointer(id, blockPeriod());
   }
 
-  function _getChallenge(uint8 pointer) internal view returns (bytes32) {
+  function getChallenge(uint8 pointer) internal view returns (bytes32) {
```

Comment:
Good one, if I remember correctly, then the main reason for me to add an `_` prefix was to prevent name clashes between a private function in a base class, and a public function in a derived class. For instance between `Proofs._startRequiringProofs()` and `TestProofs.startRequiringProofs()`. But perhaps it would be better to use virtual functions for that. Then we could drop the `_` prefix for functions alltogether.

Reply:
> can we enforce that somehow?

We can, using [solhint's `private-vars-leading-underscore` rule](https://protofire.github.io/solhint/docs/rules/naming/private-vars-leading-underscore.html), which can be used as a [hardhat plugin](https://hardhat.org/hardhat-runner/plugins/nomiclabs-hardhat-solhint) for `hardhat check`. We could then use that as a step in CI to enforce it.

Reply:
Nice, good find! We already use `solhint` in the CI, so it was easy to add this rule.
Unfortunately it means that we'll also have to prefix private variables with an `_`, but so be it.